### ISSUE TITLE
Use function symbols to validate instruction addresses in unwinders

### ISSUE THREAD

**Issue by Gankra** [luser/rust-minidump#194]
Every unwinder currently has the place to do this stubbed out, I just haven't bothered yet.

```
fn instruction_seems_valid<P>(
    instruction: Pointer,
    modules: &MinidumpModuleList,
    _symbol_provider: &P,
) -> bool
where
    P: SymbolProvider,
{
    if let Some(_module) = modules.module_at_address(instruction as u64) {
        // TODO: if mapped, check if this instruction actually maps to a function line
        true
    } else {
        false
    }
}
```

**Comment by gabrielesvelto** [comment]
I thought about doing this in the past but there's a significant downside: pointers into JIT'd code won't be recognized as valid. Until we have a good solution for that this heuristic isn't great.

**Comment by Gankra** [comment]
I'm a bit surprised to hear that since breakpad currently does exactly that for stack scanning: https://searchfox.org/mozilla-central/source/toolkit/crashreporter/google-breakpad/src/processor/stackwalker.cc#330

(but not for cfi/frame-pointers)

**Comment by gabrielesvelto** [comment]
Yeah, what happens in practice is that Breakpad will skip JIT frames so you won't see them in our stacks. If you implement this heuristic we should get identical stacks to what we get now but given they're not optimal also feel free to skip it. We can think of a long-term solution at a later stage, such as recording executable areas generated by the JIT compiler in the minidump and use those to validate pointers (and also label them in a friendlier way).

**Comment by luser** [comment]
> We can think of a long-term solution at a later stage, such as recording executable areas generated by the JIT compiler in the minidump

FYI, the `MINIDUMP_MEMORY_INFO_LIST` stream is exactly that: info about memory regions and their state.

**Comment by Gankra** [comment]
oh yeah, if we have page info that might be better

@gabrielesvelto do you have a good example of JS frames? I had been intending to investigate what happens for those.

**Comment by gabrielesvelto** [comment]
Look for crashes with `js::Call()` on the stack and you'll find plenty:

https://crash-stats.mozilla.org/search/?proto_signature=~js%3A%3ACall&_facets=signature&_sort=-date&_columns=date&_columns=signature&_columns=product&_columns=version&_columns=build_id&_columns=platform

**Comment by gabrielesvelto** [comment]
> FYI, the `MINIDUMP_MEMORY_INFO_LIST` stream is exactly that: info about memory regions and their state.

Thanks! I noticed that there's also a recently-introduced [`MINIDUMP_FUNCTION_TABLE_STREAM`](https://docs.microsoft.com/en-us/windows/win32/api/minidumpapiset/ns-minidumpapiset-minidump_function_table_stream) whose description in minidumpapiset.h is:

> Support for capturing dynamic function table state at the time of the dump.

However Microsoft's documentation is vague about it and there appears to be no documented way to write it.
